### PR TITLE
docs: add mocked mode doc to docs

### PR DESCRIPTION
**Pull request by poppyseedDev**

### CODE REVIEW

File: docs/smart_contracts/mocked.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      const unique2 = new Set(values2);
+      expect(unique2.size).to.be.greaterThanOrEqual(2);
+    }
+  });
+});
+```
+
+In the previous snippet, the first test will be run in every case, whether in testing in mocked mode (`pnpm test`) or coverage (mocked) mode (`pnpm coverage`). On the other hand, the second test will be run **only** in mocked mode (when running `pnpm test`), since snapshots only work in that specific case. The second test will be skipped if run in coverage mode, since its description string ends with `[skip-on-coverage]` and similarly, we avoid the test to fail in non-mocked mode since we check that the network name is `hardhat`.
+
+⚠️ **Warning :** Due to intrinsic limitations of the original EVM, the mocked version differ in few corner cases from the real fhEVM, the main difference is the difference in gas prices for the FHE operations.
```
Comment (jatZama) [inline docs/smart_contracts/mocked.md]: This should no longer be the case in next release fyi, so you can remove his Warning already.

File: docs/smart_contracts/mocked.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      }
+      // Expect at least two different generated values.
+      const unique2 = new Set(values2);
+      expect(unique2.size).to.be.greaterThanOrEqual(2);
+    }
+  });
+});
+```
+
+In the previous snippet, the first test will be run in every case, whether in testing in mocked mode (`pnpm test`) or coverage (mocked) mode (`pnpm coverage`). On the other hand, the second test will be run **only** in mocked mode (when running `pnpm test`), since snapshots only work in that specific case. The second test will be skipped if run in coverage mode, since its description string ends with `[skip-on-coverage]` and similarly, we avoid the test to fail in non-mocked mode since we check that the network name is `hardhat`.
```
Comment (jatZama) [inline docs/smart_contracts/mocked.md]: Please add one word "testing":
`**only** in testing mocked mode` instead of `**only** in mocked mode`

File: docs/smart_contracts/mocked.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+⚠️ **Warning :** Due to intrinsic limitations of the original EVM, the mocked version differ in few corner cases from the real fhEVM, the main difference is the difference in gas prices for the FHE operations.
+
+{% hint style="success" %}
+**Zama 5-Question Developer Survey**
+
+We want to hear from you! Take 1 minute to share your thoughts and helping us enhance our documentation and libraries. **👉** [**Click here**](https://www.zama.ai/developer-survey) to participate.
+{% endhint %}
```
Comment (yuxizama) [inline docs/smart_contracts/mocked.md]: We don't have to include the survey here

File: docs/smart_contracts/mocked.md (hunk 2 above)
Comment (yuxizama) [reply docs/smart_contracts/mocked.md]: Rewording for better readability:

In the snippet above:

- **The first test** is always run in both mocked mode (pnpm test) and coverage mode (pnpm coverage).
- **The second test** only runs in mocked mode (`pnpm test`), because snapshots only work there. It’s skipped under coverage mode if the description string ends with `[skip-on-coverage]`. We also prevent failures in non-mocked mode by checking that the network is hardhat.

File: docs/smart_contracts/mocked.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+Or equivalently:
+
+```
+npx hardhat coverage
+```
+
+Then open the file `coverage/index.html` to see the coverage results. This will increase security by pointing out missing branches not covered yet by the current test suite.
+
+**Notice:** Due to limitations in the `solidity-coverage` package, test coverage computation does not work with tests that use the `evm_snapshot` Hardhat testing method. If you are using Hardhat snapshots in your tests, we recommend adding the `[skip-on-coverage]` tag at the end of your test description. Here's an example:
```
Comment (yuxizama) [inline docs/smart_contracts/mocked.md]: {% hint style="info" %}
Due to limitations in the `solidity-coverage` package, test coverage computation does not work with tests that use the `evm_snapshot` Hardhat testing method. 
{% endhint %}

If you are using Hardhat snapshots in your tests, we recommend adding the `[skip-on-coverage]` tag at the end of your test description. Here's an example:

File: docs/smart_contracts/mocked.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+The mocked mode does **not** actually perform real encryption for encrypted types and instead runs the tests on a local Hardhat node, which is implementing the original EVM (i.e., non-fhEVM).
+
+Additionally, the mocked mode will let you use all the Hardhat-related special testing and debugging methods, such as `evm_mine`, `evm_snapshot`, `evm_revert`, etc., which are very helpful for testing.
+
+## Development Workflow with Mocked Mode
+
+When developing confidential contracts, we recommend to use first the mocked version of fhEVM for faster testing with `pnpm test` and coverage computation via `pnpm coverage`, this will lead to a better developer experience. However, keep in mind that the mocked fhEVM has some limitations and discrepancies compared to the real fhEVM node, as explained in the warning section at the end of this page.
```
Comment (yuxizama) [inline docs/smart_contracts/mocked.md]: > However, keep in mind that the mocked fhEVM has some limitations and discrepancies compared to the real fhEVM node, as explained in the warning section at the end of this page.

If the warning section will be removed, this part should be removed too?

File: docs/smart_contracts/mocked.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+## Overview
+
+**Mocked mode** is a development and testing feature provided in the `fhEVM` framework that allows developers to simulate the behavior of Fully Homomorphic Encryption (FHE) without requiring the full encryption and decryption processes to be performed. This makes development and testing cycles faster and more efficient by replacing actual cryptographic operations with mocked values, which behave similarly to encrypted data but without the computational overhead of true encryption.
+
+## Where You Can Use Mocked Mode
+
+### 1. **Hardhat Template**
+
+Mocked mode is currently supported in the [Zama Hardhat template](https://github.com/zama-ai/fhevm-hardhat-template). Developers can enable mocked mode to simulate encrypted operations while building and testing smart contracts locally. The Hardhat template includes pre-configured scripts and libraries to simplify the setup process for mocked mode.
```
Comment (yuxizama) [inline docs/smart_contracts/mocked.md:13]: Add something like: 
Refer to the [[Quick start - Hardhat]](https://docs.zama.ai/fhevm/getting-started/overview-1/hardhat) guide for instructions on using the Hardhat template.

File: docs/smart_contracts/mocked.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
+# Mocked Mode
+
+This document provides an overview of mocked mode in the fhEVM framework, explaining how it enables faster development and testing of smart contracts that use Fully Homomorphic Encryption (FHE).
+
+## Overview
+
+**Mocked mode** is a development and testing feature provided in the `fhEVM` framework that allows developers to simulate the behavior of Fully Homomorphic Encryption (FHE) without requiring the full encryption and decryption processes to be performed. This makes development and testing cycles faster and more efficient by replacing actual cryptographic operations with mocked values, which behave similarly to encrypted data but without the computational overhead of true encryption.
+
+## Where You Can Use Mocked Mode
```
Comment (yuxizama) [inline docs/smart_contracts/mocked.md]:
```suggestion
## How to use mocked mode
```

File: docs/smart_contracts/mocked.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+### 1. **Hardhat Template**
+
+Mocked mode is currently supported in the [Zama Hardhat template](https://github.com/zama-ai/fhevm-hardhat-template). Developers can enable mocked mode to simulate encrypted operations while building and testing smart contracts locally. The Hardhat template includes pre-configured scripts and libraries to simplify the setup process for mocked mode.
+
+### 2. **Foundry (Coming Soon)**
+
+Mocked mode support is planned for [Foundry](./write_contract/foundry.md) in future releases.
+
+## How Mocked Mode Works
```
Comment (yuxizama) [inline docs/smart_contracts/mocked.md]:
```suggestion
## How mocked mode works
```

File: docs/smart_contracts/mocked.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+**Mocked mode** is a development and testing feature provided in the `fhEVM` framework that allows developers to simulate the behavior of Fully Homomorphic Encryption (FHE) without requiring the full encryption and decryption processes to be performed. This makes development and testing cycles faster and more efficient by replacing actual cryptographic operations with mocked values, which behave similarly to encrypted data but without the computational overhead of true encryption.
+
+## Where You Can Use Mocked Mode
+
+### 1. **Hardhat Template**
+
+Mocked mode is currently supported in the [Zama Hardhat template](https://github.com/zama-ai/fhevm-hardhat-template). Developers can enable mocked mode to simulate encrypted operations while building and testing smart contracts locally. The Hardhat template includes pre-configured scripts and libraries to simplify the setup process for mocked mode.
+
+### 2. **Foundry (Coming Soon)**
```
Comment (yuxizama) [inline docs/smart_contracts/mocked.md]:
```suggestion
### 2. **Foundry (Coming soon)**
```

File: docs/smart_contracts/mocked.md [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+This document provides an overview of mocked mode in the fhEVM framework, explaining how it enables faster development and testing of smart contracts that use Fully Homomorphic Encryption (FHE).
+
+## Overview
+
+**Mocked mode** is a development and testing feature provided in the `fhEVM` framework that allows developers to simulate the behavior of Fully Homomorphic Encryption (FHE) without requiring the full encryption and decryption processes to be performed. This makes development and testing cycles faster and more efficient by replacing actual cryptographic operations with mocked values, which behave similarly to encrypted data but without the computational overhead of true encryption.
+
+## Where You Can Use Mocked Mode
+
+### 1. **Hardhat Template**
```
Comment (yuxizama) [inline docs/smart_contracts/mocked.md]:
```suggestion
### 1. **Hardhat template**
```

File: docs/smart_contracts/mocked.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+## How Mocked Mode Works
+
+For faster testing iterations, instead of launching all the tests on the local fhEVM node, which could last several minutes, you can use a mocked version of the fhEVM by running `pnpm test`. The same tests should (almost always) pass as is, without any modification; neither the JavaScript files nor the Solidity files need to be changed between the mocked and the real version.
+
+The mocked mode does **not** actually perform real encryption for encrypted types and instead runs the tests on a local Hardhat node, which is implementing the original EVM (i.e., non-fhEVM).
+
+Additionally, the mocked mode will let you use all the Hardhat-related special testing and debugging methods, such as `evm_mine`, `evm_snapshot`, `evm_revert`, etc., which are very helpful for testing.
+
+## Development Workflow with Mocked Mode
```
Comment (yuxizama) [inline docs/smart_contracts/mocked.md]:
```suggestion
## Development workflow with mocked mode
```

**Review by yuxizama — COMMENTED** [review]
Thank you @poppyseedDev It's good! I added a few suggestions. 
Besides, could you also add a sentence in the Quick Start - 3. Testing in mocked mode linking to this doc?

File: docs/smart_contracts/mocked.md (hunk 2 above)
Comment (jatZama) [reply docs/smart_contracts/mocked.md:139]: You missed my point entirely here, you missed again my key remark. Since you skipped again the key word to ADD : ""testing"/ please reconsider my comment?

Comment (jatZama) [reply docs/smart_contracts/mocked.md]: sorry my bad i was looking at an old commit. looks good to me

File: docs/smart_contracts/mocked.md [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+      }
+      // Expect at least two different generated values.
+      const unique2 = new Set(values2);
+      expect(unique2.size).to.be.greaterThanOrEqual(2);
+    }
+  });
+});
+```
+
+In the previous snippet, the first test will be run in every case, whether in testing in mocked mode (`pnpm test`) or coverage (mocked) mode (`pnpm coverage`). On the other hand, the second test will be run **only** in mocked mode (when running `pnpm test`), since snapshots only work in that specific case. The second test will be skipped if run in coverage mode, since its description string ends with `[skip-on-coverage]` and similarly, we avoid the test to fail in non-mocked mode since we check that the network name is `hardhat`.
```
Comment (jatZama) [inline docs/smart_contracts/mocked.md]: Can you remove this line entirely please? I think it is redundant and less clear with the lines that come after. Otherwise it looks good.

**Review by jatZama — APPROVED** [review]
All good now.
